usr/src/uts/common/fs/zfs/zap_micro.c

```diff
@@ -377,9 +379,13 @@ mzap_open(objset_t *os, uint64_t obj, dmu_buf_t *db)
 	zap->zap_object = obj;
 	zap->zap_dbuf = db;
 
-	if (*(uint64_t *)db->db_data != ZBT_MICRO) {
+	if (zap_block_type != ZBT_MICRO) {
 		mutex_init(&zap->zap_f.zap_num_entries_mtx, 0, 0, 0);
 		zap->zap_f.zap_block_shift = highbit64(db->db_size) - 1;
+		if (zap_block_type != ZBT_HEADER  || zap_magic != ZAP_MAGIC) {
```

Comment:
Yeah, fixed and force-pushed.  I guess zettabot will do, but be careful, it still may crash at boot time.